core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1071,53 +1298,107 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
          throw new PersistenceException("Could not add new merged entry", e);
       }
    }
+
    @Override
-   public void purge(Executor threadPool, final PurgeListener task) {
+   public Publisher<MarshallableEntry<K, V>> purgeExpired() {
       long now = timeService.wallClockTime();
-      List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<>();
-      synchronized (entries) {
-         for (Iterator<Map.Entry<K, FileEntry>> it = entries.entrySet().iterator(); it.hasNext(); ) {
-            Map.Entry<K, FileEntry> next = it.next();
-            FileEntry fe = next.getValue();
-            if (fe.isExpired(now)) {
-               it.remove();
-               entriesToPurge.add(new KeyValuePair<>(next.getKey(), fe));
-            }
-         }
-      }
+      UnicastProcessor<MarshallableEntry<K, V>> processor = UnicastProcessor.create();
+      blockingManager.runBlocking(() -> {
+         blockingPurgeExpired(now, processor);
+      }, "sfs-purgeExpired");
+      return Flowable.fromPublisher(processor.serialize());
```

Comment:
Probably not.
I think at the time I wanted to publish from multiple threads, but I gave up on that (and I should have probably passed `processor.serialize()` to `blockingPurgeExpired()` instead of returning it to the caller, anyway.

Reply:
How were you thinking of using multiple threads here? If anything we could add multi threading to the PersistenceManagerImpl to purge stores in parallel maybe, but I see no reason for the SFS to do multi threading here.

Reply:
I was thinking of using one thread to iterate over each segment, all of them publishing entries on the same serialized processor.